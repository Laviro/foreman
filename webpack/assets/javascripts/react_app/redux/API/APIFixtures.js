import Immutable from 'seamless-immutable';
import { noop } from '../../common/helpers';

export const key = 'SOME_KEY';
export const APIRequest = noop;
export const polling = 3000;
export const initialState = Immutable({
  polling: {},
});
export const stateWithKey = Immutable({
  polling: {
    [key]: 1,
  },
});
