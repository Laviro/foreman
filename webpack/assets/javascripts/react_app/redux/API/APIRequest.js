import { API } from './';
import { actionTypeGenerator } from './APIActionTypeGenerator';
import { noop } from '../../common/helpers';

export const get = async (
  {
    key,
    url,
    headers = {},
    params = {},
    actionTypes = {},
    handleError = noop,
    payload = {},
  },
  { dispatch }
) => {
  const { REQUEST, SUCCESS, FAILURE } = actionTypeGenerator(key, actionTypes);
  const modifiedPayload = { ...payload, url };
  dispatch({
    type: REQUEST,
    key,
    payload: modifiedPayload,
  });
  try {
    const { data } = await API.get(url, headers, params);
    dispatch({
      type: SUCCESS,
      key,
      payload: modifiedPayload,
      response: data,
    });
  } catch (error) {
    handleError(error);
    dispatch({
      type: FAILURE,
      key,
      payload: modifiedPayload,
      response: error,
    });
  }
};
