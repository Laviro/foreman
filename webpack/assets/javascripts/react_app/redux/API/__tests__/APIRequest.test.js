import { API } from '../';
import IntegrationTestHelper from '../../../common/IntegrationTestHelper';
import { action } from '../APIFixtures';
import { apiRequest } from '../APIRequest';

const data = { results: [1] };
jest.mock('../');

describe('API get', () => {
  const store = { dispatch: jest.fn() };
  beforeEach(() => {
    store.dispatch = jest.fn();
  });

  it('should dispatch request and success actions on resolve', async () => {
    const apiSuccessResponse = { data };
    API.get.mockImplementation(
      () =>
        new Promise((resolve, reject) => {
          resolve(apiSuccessResponse);
        })
    );
    const modifiedAction = { ...action };
    modifiedAction.payload.handleSuccess = jest.fn();
    apiRequest(modifiedAction, store);
    await IntegrationTestHelper.flushAllPromises();
    expect(modifiedAction.payload.handleSuccess).toHaveBeenLastCalledWith(
      apiSuccessResponse
    );
    expect(store.dispatch.mock.calls).toMatchSnapshot();
  });

  it('should dispatch request and failure actions on reject', async () => {
    const apiError = new Error('bad request');
    API.get.mockImplementation(
      () =>
        new Promise((resolve, reject) => {
          reject(apiError);
        })
    );
    const modifiedAction = { ...action };
    modifiedAction.payload.handleError = jest.fn();
    apiRequest(modifiedAction, store);
    await IntegrationTestHelper.flushAllPromises();
    expect(modifiedAction.payload.handleError).toHaveBeenLastCalledWith(
      apiError
    );
    expect(store.dispatch.mock.calls).toMatchSnapshot();
  });
});
