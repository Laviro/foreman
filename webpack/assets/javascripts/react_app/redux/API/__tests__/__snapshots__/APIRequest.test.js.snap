// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API get should dispatch request and failure actions on reject 1`] = `
Array [
  Array [
    [Error: bad request],
    [Function],
  ],
]
`;

exports[`API get should dispatch request and failure actions on reject 2`] = `
Array [
  Array [
    Object {
      "key": "SOME_KEY",
      "payload": Object {
        "id": 2,
        "url": "some/url",
      },
      "type": "SOME_KEY_REQUEST",
    },
  ],
  Array [
    Object {
      "key": "SOME_KEY",
      "payload": Object {
        "id": 2,
        "url": "some/url",
      },
      "response": [Error: bad request],
      "type": "SOME_KEY_FAILURE",
    },
  ],
]
`;

exports[`API get should dispatch request and success actions on resolve 1`] = `
Array [
  Array [
    Object {
      "key": "SOME_KEY",
      "payload": Object {
        "id": 2,
        "url": "some/url",
      },
      "type": "SOME_KEY_REQUEST",
    },
  ],
  Array [
    Object {
      "key": "SOME_KEY",
      "payload": Object {
        "id": 2,
        "url": "some/url",
      },
      "response": Object {
        "results": Array [
          1,
        ],
      },
      "type": "SOME_KEY_SUCCESS",
    },
  ],
]
`;

exports[`API get should dispatch stop interval on API error 1`] = `
Array [
  Array [
    [Error: bad request],
    [Function],
  ],
]
`;

exports[`API get should dispatch stop interval on API error 2`] = `
Array [
  Array [
    Object {
      "key": "SOME_KEY",
      "payload": Object {
        "id": 2,
        "url": "some/url",
      },
      "type": "SOME_KEY_REQUEST",
    },
  ],
  Array [
    Object {
      "key": "SOME_KEY",
      "payload": Object {
        "id": 2,
        "url": "some/url",
      },
      "response": [Error: bad request],
      "type": "SOME_KEY_FAILURE",
    },
  ],
]
`;
